src/G2/Execution/Reducer.hs

```diff
@@ -649,465 +619,297 @@ instance (Named a, Named b) => Named (C a b) where
--- Discards it if in SWHNF and true_assert is false
-data AcceptIfViolatedHalter = AcceptIfViolatedHalter
-
-instance Halter AcceptIfViolatedHalter () t where
-    initHalt _ _ = ()
-    updatePerStateHalt _ _ _ _ = ()
-    stopRed _ _ _ s =
-        case isExecValueForm s of
-            True 
-                | true_assert s -> return Accept
-                | otherwise -> return Discard
-            False -> return Continue
-    stepHalter _ _ _ _ _ = ()
+            , updatePerStateHalt = \(C hv1 hv2) proc s ->
+                let
+                    hv1' = updatePerStateHalt h1 hv1 proc s
+                    hv2' = updatePerStateHalt h2 hv2 proc s
+                in
+                C hv1' hv2'
 
--- | Allows execution to continue until the step counter hits 0, then discards the state
-data ZeroHalter = ZeroHalter Int
+            , discardOnStart = \(C hv1 hv2) proc s ->
+                let
+                    b1 = discardOnStart h1 hv1 proc s
+                    b2 = discardOnStart h2 hv2 proc s
+                in
+                b1 || b2
 
-instance Halter ZeroHalter Int t where
-    initHalt (ZeroHalter n) _ = n
-    updatePerStateHalt _ hv _ _ = hv
-    stopRed = halterIsZero
-    stepHalter = halterSub1
+            , stopRed = \(C hv1 hv2) proc s -> do
+                hc1 <- stopRed h1 hv1 proc s
+                hc2 <- stopRed h2 hv2 proc s
 
-halterSub1 :: Halter h Int t => h -> Int -> Processed (State t) -> [State t] -> State t -> Int
-halterSub1 _ h _ _ _ = h - 1
+                return $ min hc1 hc2
 
-halterIsZero :: Halter h Int t => h -> Int -> Processed (State t) -> State t -> IO HaltC
-halterIsZero _ 0 _ _ = return Discard
-halterIsZero _ _ _ _ = return Continue
+            , stepHalter = \(C hv1 hv2) proc xs s ->
+                let
+                    hv1' = stepHalter h1 hv1 proc xs s
+                    hv2' = stepHalter h2 hv2 proc xs s
+                in
+                C hv1' hv2'
 
-data MaxOutputsHalter = MaxOutputsHalter (Maybe Int)
+            , updateHalterWithAll = \shv ->
+                let
+                    shv1 = map (\(s, C hv1 _) -> (s, hv1)) shv
+                    shv2 = map (\(s, C _ hv2) -> (s, hv2)) shv
 
-instance Halter MaxOutputsHalter (Maybe Int) t where
-    initHalt (MaxOutputsHalter m) _ = m
-    updatePerStateHalt _ hv _ _ = hv
-    stopRed _ m (Processed {accepted = acc}) _ =
-        case m of
-            Just m' -> return $ if length acc >= m' then Discard else Continue
-            _ -> return Continue
-    stepHalter _ hv _ _ _ = hv
+                    shv1' = updateHalterWithAll h1 shv1
+                    shv2' = updateHalterWithAll h2 shv2
+                in
+                map (uncurry C) $ zip shv1' shv2'
+            }
+{-# INLINE (<~>) #-}
+
+{-# INLINE swhnfHalter #-}
+swhnfHalter :: Monad m => Halter m () t
+swhnfHalter = mkStoppingHalter stop
+    where
+        stop _ _ s =
+            case isExecValueForm s of
```

Comment:
I vaguely remember the two of us having a conversation at some point about why SWHNF and exec-value form are different things that shouldn't be conflated.  Maybe I'm misremembering, but should we change the name of this halter just to be thorough?

Reply:
I also only vaguely remember this conversation.  But my guess as to the source of the confusion is: there's kind of an unfortunate conflation of names here, because the Nebula paper _does not_ consider a symbolic variable with an ADT type to be in SWHNF, but the PLDI'19 G2 paper _does_.
https://billhallahan.github.io/files/lazy-cf-symex.pdf (see end of Section 3.2)
It would be good to sort this out in the codebase and make it consistent, but it's not currently clear to me how best to do that (SWHNF1 and SWHNF2 seems messy.) Especially since I'm just adopting the current name (SWHNFHalter datatype -> swhnfHalter function) I might just advocate leaving this for now.